### PR TITLE
Bind expr fixes

### LINKED ISSUE #509: bind_expr Bugs
1. bind_expr doesn't accept an argument that is a curosr to an expressions, but rather it must be given a list always if we are using a cursor. This is a bug in the arguments processor. 
Bug example:
```
def test_bind_expr_issue():
    @proc
    def bar(n: size, x: f32[n], y: f32[n], z: f32[n]):
        for i in seq(0, n):
            x[i] = y[i]
            x[i] = z[i]
    
    yi_cursor = bar.find("y[_]")
    zi_cursor = bar.find("z[_]")
    bar = bind_expr(bar, [yi_cursor], "y_read")     <----  Okay
    bar = bind_expr(bar, zi_cursor, "z_read")       <----- Error
````

2. Passing a list of more than one cursor to bind_expr binds only the first cursor unless `cse` is set to True. I find this very confusing because why else would I want to pass a list of many cursors:
```
def test_bind_expr_issue():
    @proc
    def bar(n: size, x: f32[n], y: f32[n], z: f32[n]):
        for i in seq(0, n):
            x[i] = y[i]
            z[i] = y[i]
    
    yi_cursors = bar.find("y[_]", many=True)
    print(bind_expr(bar, yi_cursors, "y_read"))
    print(bind_expr(bar, yi_cursors, "y_read", cse=True))
```
```
def bar(n: size, x: f32[n] @ DRAM, y: f32[n] @ DRAM, z: f32[n] @ DRAM):
    for i in seq(0, n):
        y_read: R
        y_read = y[i]
        x[i] = y_read
        z[i] = y[i]
def bar(n: size, x: f32[n] @ DRAM, y: f32[n] @ DRAM, z: f32[n] @ DRAM):
    for i in seq(0, n):
        y_read: R
        y_read = y[i]
        x[i] = y_read
        z[i] = y_read
```
Perhpas we should internally set cse=True whenever the user passes a list of cursors. I think generally it is unclear to me what cse implies in the context of cursors and in the contex of template arguments. We should at least document this behvaior if it is what is expected.
3. bind_expr currently assigns the type R to the new allocation. I think this is a bit confusing since when I bind something I would expect the default to be the type of the original expressions, otherwise you can easily get into a situation where you have cross-precision computation even though your original computation didn't have that. Changing the type of some computation shouldn't be something that is silently done by a scheduling operations, but only by explicit direction from the processor by using set_precision. Another reason why this is confusing is that stage_mem would use the precision of the staged buffer which makes it a bit inconsinstent for bind_expr to do something else.

### PR DESCRIPTION
**Pull request by SamirDroubi**
Partially addresss #540 and fixes #509 .

### CODE REVIEW

**Comment by SamirDroubi** [comment]
It is expected that BLAS tests will fail.

File: apps/aarch64/filter1D/filter1D_cursor_forwarding.py [hunk 1]
```diff
@@ -42,7 +42,6 @@ def filter1D(ow: size, kw: size, x: f32[ow + kw - 1], y: f32[ow], w: f32[kw]):
 # set memories & precision
 filter1D = set_memory(filter1D, "sum #1", Neon)
 filter1D = set_memory(filter1D, "xX4", Neon)
-filter1D = set_precision(filter1D, "xX4", "f32")
```
Comment (yamaguchi1024) [inline apps/aarch64/filter1D/filter1D_cursor_forwarding.py:45]: Why was this removed?

File: tests/test_apps.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def test_gemmini_matmul(golden):
     assert _test_app(module_file) == golden
 
 
-@pytest.mark.slow
+@pytest.mark.skip(reason="")
```
Comment (yamaguchi1024) [inline tests/test_apps.py:47]: Gemmini conv shouldn't be skipped

Comment (SamirDroubi) [reply tests/test_apps.py]: My bad, I usually disable this test because it takes forever to run locally. Accidently, pushed that code.

File: apps/aarch64/filter1D/filter1D_cursor_forwarding.py (hunk 1 above)
Comment (SamirDroubi) [reply apps/aarch64/filter1D/filter1D_cursor_forwarding.py:45]: Since bind_expr automatically infers the type it is no longer necessary to set the precision.

File: tests/test_apps.py (hunk 2 above)
Comment (yamaguchi1024) [reply tests/test_apps.py]: I think pytest took --fast option or sth like that to skip slow tests from the commandline

File: apps/aarch64/filter1D/filter1D_cursor_forwarding.py (hunk 1 above)
Comment (yamaguchi1024) [reply apps/aarch64/filter1D/filter1D_cursor_forwarding.py:45]: I see, nice!

**Review by yamaguchi1024 — APPROVED** [review]
